staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -227,12 +199,37 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
 	}
 	for {
 		origState, err := s.getState(getResp, key, v, false)
-		if err != nil {
+		if err != nil && !storage.IsTransformerError(err) {
 			return err
 		}
-		if err := preconditions.Check(key, origState.obj); err != nil {
+		if err != nil && storage.IsTransformerError(err) && preconditions != nil {
+			// Checking preconditions requires origState.obj, which
+			// is not set in case of TransformerError.
+			// conditionalDelete fails closed in this case. If the
+			// user still wants to delete the object, they can
+			// retry the delete with empty preconditions.
 			return err
 		}
+		// We let conditionalDelete skip validateDeletion, delete the
```

Comment:
Is it better to skip validateDeletion or call it with a nil object? Currently, delete admission is always called, regardless of transform error, with a nil oldObject. would it be preferable to continue calling delete admission with a nil oldObject in the case of a transform error over skipping it entirely?

Reply:
I can't decide which bothers me more, giving an inconsistent API to admission consumers (*sometimes* getting a nil oldObject in a delete admission call), skipping delete admission entirely, or requiring talking to etcd to delete a malformed value.

Will CRD webhook conversion errors and KMS transform errors exercise this path? I'm not sure I'd want to abandon calling admission if an external system like that flakes.

Reply:
Indeed the KMS transform is part of the transformer stack: https://github.com/kubernetes/kubernetes/blob/4d3d153210c0788b8e188c467d31061470d8c042/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L190.

I'll trace the CRD webhook path later.

Reply:
Among all the options, "giving an inconsistent API to admission consumers" seems to be the lesser of the evils.

I added the last commit to send nil objects to `validateDelete`.

Reply:
Discussed offline. since we already have to look up the existing object in the REST handler to verify there are no remaining finalizers, a persistent transform error is already blocking, and we'll never make it here. Continuing to simply return if we encounter a transform error (rather than skipping admission or passing a nil object) is no worse than the current state, and lets us give a consistent API to admission authors.

Reply:
Let's drop the transformation error changes from this PR to keep it focused on passing existing objects to delete admission